### DISCUSSION TITLE
How should migration be done to Production?

### DISCUSSION THREAD

**Discussion by TommyLeong** [prisma/prisma#24571] [Q&A]
### Question

Hello everyone, I'm here to clarify my understanding and actual practise everyone does in the Prod environment.

First of all, I have read this [documentation here](https://www.prisma.io/docs/orm/prisma-migrate/workflows/development-and-production#production-and-testing-environments) before asking. Im fully aware this is the recommended command `npx prisma migrate deploy` to be used in Production. 

## Scenario 1
My understanding, which is also what I'm practising now.
1. During implementation phases, I amend the DB directly without using `schema.prisma`
2. To keep my `schema.prisma` up to date, I tend to execute `prisma db pull` follow by `prisma generate` too.

When it comes to Production, 
1. With the "latest" `schema.prisma` I have on hand, I need to have them reflected on my Production DB.
2. Locally, I execute `npx prisma migrate dev --create-only` to form the `migration.sql`, without applying it.
3. Locally, I execute `npx prisma migrate deploy` to apply these changes.

Am I on the right track? 

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
// Add your schema.prisma
```

```ts
// Add any relevant Prisma Client queries here
```

- OS: 
- Database: 
- Node.js version: 

- Run `prisma -v` to see your Prisma version and paste it

**Comment by ludralph** [comment]
Hi @TommyLeong 👋 

Thank you for raising this question.

Your understanding is almost correct but there are a few adjustments needed to align with the recommended Prisma workflows for production environments.

**Scenario 1: Development Phase**

- Amending the DB directly: While you can amend the database directly, it's not the recommended approach. Instead, you should make changes to your `schema.prisma` file and then generate migrations.

- Keeping `schema.prisma` up to date: Using `prisma db pull` to update your `schema.prisma` file is correct if you have made changes directly to the database. However, this should be followed by creating a migration to capture these changes.

**Recommended Development Workflow**

- Make Changes to schema.prisma: Update your `schema.prisma` file with the desired changes.

- Generate and Apply Migrations: Use the `prisma migrate dev` command to generate and apply migrations during development with `npx prisma migrate dev`.

**Scenario 1: Production Phase**

Reflecting Changes in Production DB: You should not use `prisma migrate dev` in production. Instead, use `prisma migrate deploy` to apply migrations in a production environment.

**Recommended Production Workflow**

- Production and Testing Environments: In [production and testing](https://www.prisma.io/docs/orm/prisma-migrate/workflows/development-and-production#production-and-testing-environments) environments, use the `migrate deploy` command to apply migrations. This command should generally be part of an automated CI/CD pipeline, and it is not recommended to run this command locally to deploy changes to a production database. See this [section](https://www.prisma.io/docs/orm/prisma-client/deployment/deploy-migrations-from-a-local-environment) of the documentation.

-  Automated CI/CD Pipeline: Setting up an automated CI/CD pipeline is highly recommended. The pipeline should handle deployment to staging and production environments and use migrate deploy in a pipeline step.

- Squashing Migrations: [Squashing migrations](https://www.prisma.io/docs/orm/prisma-migrate/workflows/squashing-migrations#creating-a-clean-history-in-a-production-environment) can be useful when developing with a branch-based workflow. During a large local development effort on a feature branch, you might generate multiple migrations using `migrate dev`. After the feature is finished, the migration history might contain unnecessary intermediate steps that are unwanted in the final migration history that will be pushed to the main branch.

Do let us know if you have further questions.

**Reply by TommyLeong** [reply]
Hello @ludralph , thank you for the reply. It's very helpful with your details sharing.

I have a follow up question on the Production workflow, Im aware `prisma migrate dev` is not a recommended way to use in Production. But without running it with the flag `--create-only`, how is the migration.sql getting form? Because my understanding is `prisma migrate deploy` would need something to migrate deploy, which in this case the migration.sql

**Reply by ludralph** [reply]
Hi @TommyLeong 👋 

Apologies for the late reply here.

In your development environment, you use `prisma migrate dev --create-only` to create a migration file without applying it. This command generates the `migration.sql` file based on the changes in your `schema.prisma`. You can then review and customize the generated SQL if needed.

After running the above command, you can find the generated SQL file in the `prisma/migrations` directory. You can edit this file to make any necessary changes. Once you are satisfied with the migration, you can apply it in your development environment using:
```
npx prisma migrate dev
```
After testing the migration in your development environment, you can deploy it to your production environment using:
```
npx prisma migrate deploy
```
The `prisma migrate deploy` command will look for any new migration files in the `prisma/migrations` directory and apply them to your production database.

The prisma migrate deploy command uses the following files:

1. Migration Files:

The SQL migration files located in the `prisma/migrations` directory. These files represent the history of changes made to your database schema. The `prisma migrate deploy` command applies all pending migrations found in this directory.

2. Prisma Schema:

The `schema.prisma` file, which defines the structure of your database schema. While prisma migrate deploy primarily uses the migration files to apply changes, it reads the schema.prisma file to get the url and provider fields.

You should commit the contents of the `.prisma/migrations` folder, including the `migration_lock.toml` file, and the `schema.prisma` file to source control. This ensures that your migration history is preserved and can be deployed to other environments.

**Reply by grayhams74s** [reply]
Hello Can you Help me? PS C:\Users\Jun\Desktop\versions\gymsoftware> npx prisma migrate deploy
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma

No migration found in prisma/migrations

Error: P3005

The database schema is not empty. Read more about how to baseline an existing production database: https://pris.ly/d/migrate-baseline

Im getting this kind of error when I try npx migrate deploy

**Reply by ludralph** [reply]
Hi @grayhams74s 

This error occurs because Prisma Migrate expects an empty database when you're running migrations for the first time.
To resolve this issue, you need to baseline your database. Baselining is the process of initializing a migration history for a database that already exists and contains data that must be maintained.

1. Create an initial migration:
```
mkdir -p prisma/migrations/0_init
npx prisma migrate diff --from-empty --to-schema-datamodel prisma/schema.prisma --script > prisma/migrations/0_init/migration.sql
```

2. Review the generated migration file to ensure it correctly represents your current database schema.

3. Mark the migration as applied:
```
npx prisma migrate resolve --applied 0_init
```
After completing these steps, you should be able to run `npx prisma migrate deploy` without encountering the error.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!;

**Comment by mrksph** [comment]
Hi @ludralph, I’m running Postgres + a Next.js/Prisma app in Dokploy on my homelab. If migrations are run via GitHub Actions, I assume the runner needs direct DB connectivity. In a homelab setup, is the recommended approach to use a self-hosted runner (or run migrations on the server) rather than exposing Postgres publicly?
